mono/metadata/mono-perfcounters.c

```diff
@@ -1591,98 +1604,104 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
 }
 
 /* this is an icall */
-MonoArray*
-mono_perfcounter_category_names (void)
+MonoArrayHandle
+mono_perfcounter_category_names (MonoError *error)
 {
-	ERROR_DECL (error);
 	int i;
-	MonoArray *res;
+	MonoArrayHandle res;
 	MonoDomain *domain = mono_domain_get ();
 	GSList *custom_categories, *tmp;
+	MonoStringHandle name = MONO_HANDLE_NEW (MonoString, NULL);
+
 	perfctr_lock ();
+
 	custom_categories = get_custom_categories ();
-	res = mono_array_new_checked (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+	res = mono_array_new_handle (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+
 	if (!is_ok (error)) {
-		res = NULL;
+		res = NULL_HANDLE_ARRAY;
 		goto leave;
 	}
 
 	for (i = 0; i < NUM_CATEGORIES; ++i) {
 		const CategoryDesc *cdesc = &predef_categories [i];
-		MonoString *name = mono_string_new_checked (domain, cdesc->name, error);
+		mono_string_new_utf8_assign (name, domain, cdesc->name, cdesc->name_length, error);
```

Comment:
Could we use `HANDLE_LOOP_PREPARE` and make local frames instead of adding the `_assign` operations?

(although have to be careful about `goto_if_nok`)

Reply:
It is possible. It is not my preference. The result is larger (even if not run) and probably slower (may not matter, but feels hypocritical esp. when managing performance counters), and possibly harder to reason about.

The `HANDLE_LOOP_PREPARE` step is optional actually -- it is an optimization, as the code can access the global, null, and then later get the TLS (in loop, granted, slower).

The real problem imho is placing the begin/end macros carefully, and minding the continue/break/return, as you indicate.

Reply:
This also feels contradictory to the support afforded to arrays.
There the pattern is very much to create handle outside of loop, and get elements into it.
All types tend to be used in loops, not just arrays.

That said, I tried it out and it wasn't difficult, for me.
Assignment/reuse is a natural idiom. Granted, at odds with functional programming.

Reply:
https://github.com/mono/mono/pull/12276 shows the other way. I still prefer to avoid the push/pop/allocate in loop.